[GSB] Improve handling of layout constraints

swiftix opened the pull request:
This PR addresses TODOs from #8241.

- It supports merging for layout constraints, e.g., if both a _Trivial constraint and a _Trivial(64) constraint appear on a type parameter, we keep only _Trivial(64) as a more specific layout constraint. We do a similar thing for ref-counted/native-ref-counted. The overall idea is to keep the more specific of two compatible layout constraints.
- The presence of a superclass constraint implies a layout constraint, e.g., a superclass constraint implies _Class or _NativeClass layout constraint.

swiftix committed:
Extend the representation of layout constraints

- Add support for _Class and _NativeClass layout constraints, which are supposed to represent T: Superclass and P: class constraints in the future.
- Use the re-factoring to also reduce the number of dynamic allocations when creating layout constraints. Simple non-parametrized layout constraints are now represented as statically allocated singletons (static members of LayoutConstraintInfo).

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@DougGregor Please have a look.

DougGregor commented on lib/AST/ASTContext.cpp:
```diff
@@ -4074,6 +4074,31 @@ LayoutConstraint LayoutConstraint::getLayoutConstraint(LayoutConstraintKind Kind
+    case LayoutConstraintKind::UnknownLayout:
+      return LayoutConstraint(
+          &LayoutConstraintInfo::UnknownLayoutConstraintInfo);
+    default:
```
We don't need the default, right? It'll end up suppressing warnings that we'd rather see when the switch isn't covering all of the cases.

swiftix replied:
Fixed.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -106,6 +110,24 @@ bool LayoutConstraintInfo::isNativeRefCountedObject(LayoutConstraintKind Kind) {
 
+bool LayoutConstraintInfo::isClass(LayoutConstraintKind Kind) {
+  return Kind == LayoutConstraintKind::Class;
+}
```
Should this return true when `kind == LayoutConstraintKind::NativeClass`? It "is" a class constraint, after all.

swiftix replied:
For this I introduced `isAnyClass`.  This predicate is intended to be used to check the exact kind.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -114,4 +136,61 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+  }
+
+  if ((*this)->isClass() && Other->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
```
Missing a `return Other;` here

swiftix replied:
Oops. I forgot it while copy&pasting the code.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -114,4 +136,61 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+
+  if (Other->isClass() && (*this)->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
+  }
```
Missing a `return *this` here

swiftix replied:
Same here.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -114,4 +136,61 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+    // Prefer a more concrete _NativeClass conttraint.
+  }
+
+  return LayoutConstraint::getUnknownLayout();
```
Might it be better to have a 2-D table to describe the results? It's pretty hard to make sure that this long sequence of `if` statements does the right thing.

swiftix replied:
@DougGregor Please have a look at the update that I pushed. It uses a 2-D table, but I'm not quite happy with it. The initialization of the table looked pretty ugly. I had to use macros to shorten it. And I'm still not quite convinced it is really easier. WDYT?

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -106,6 +110,24 @@ bool LayoutConstraintInfo::isNativeRefCountedObject(LayoutConstraintKind Kind) {
+
+bool LayoutConstraintInfo::isRefCounted(LayoutConstraintKind Kind) {
+  return Kind == LayoutConstraintKind::RefCountedObject ||
+         Kind == LayoutConstraintKind::Class;
```
Shouldn't this also check for `Kind == LayoutConstraintKind::NativeRefCountedObject` and `Kind == LayoutConstraintKind::NativeClass`?

DougGregor commented on test/Generics/superclass_constraint.swift:
```diff
@@ -100,6 +102,7 @@ class C2 : C, P4 { }
 // CHECK: superclassConformance3
 // CHECK: Requirements:
 // CHECK-NEXT: τ_0_0 : C2 [τ_0_0: Explicit @ {{.*}}:46]
+// CHECK-NEXT: τ_0_0 : _NativeClass [τ_0_0: Explicit @ {{.*}}:46 -> Superclass]
```
Yay.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -114,4 +143,74 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+  if (!initializedMergeTable) {
+    // initialize the merge table once.
+    initializedMergeTable = true;
+    MERGE_TO(TrivialOfAtMostSize, Trivial, TrivialOfAtMostSize);
```
You should be able to use initializer list nesting `{ { ... }, { ... }, ... }` to avoid the imperative initialization and `initializedMergeTable` bit. It'll also make sure that cases never get skipped, e.g., the missing `merge(TrivialOfExactSize, TrivialOfExactSize)` case.

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -114,4 +143,74 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+    return LayoutConstraint::getUnknownLayout();
+
+  if ((*this)->getKind() == mergeKind)
+    return *this;
```
When the result is one of the sized cases (`TrivialOfExactSize` or `TrivialOfAtMostSize`), you'll need to do some checking of the size to determine the proper result. The lookup table doesn't always suffice in these cases... for example, the `merge(TrivialOfExactSize, TrivialOfExactSize)` case might produce `TrivialOfExactSize` (if the sizes are the same) and might produce `TrivialOfAtMostSize` or `UnknownLayout` (if the sizes are different; we can choose which semantics we prefer).

swiftix replied:
Done.

DougGregor commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -170,6 +196,18 @@ class LayoutConstraintInfo : public llvm::FoldingSetNode {
 
   static bool isNativeRefCountedObject(LayoutConstraintKind Kind);
 
+  static bool isAnyRefCountedObject(LayoutConstraintKind Kind);
```
I don't see the reason for these `static` versions. They just seem to bloat the API.

DougGregor commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -108,6 +114,26 @@ class LayoutConstraintInfo : public llvm::FoldingSetNode {
+    return isNativeClass(Kind);
+  }
+
+  bool isAnyClass() const {
```
I still find it confusing that we have `isClass()` and `isAnyClass()`. If someone wants the `isClass()` behavior, they can just check the kind exactly. Are you opposed to giving `isClass`() the behavior that `isAnyClass()` has (i.e., so `isClass()` answers the question "is it some kind of class?") and removing `isAnyClass()` entirely?

swiftix replied:
Done.

DougGregor commented on include/swift/AST/LayoutConstraint.h:
```diff
@@ -108,6 +114,26 @@ class LayoutConstraintInfo : public llvm::FoldingSetNode {
+    return isAnyClass(Kind);
+  }
+
+  bool isRefCounted() const {
```
These behavior the way I'd expect: `isRefCounted()` is `true` for any layout kind that ensures reference-counting.

swiftix commented:
@DougGregor I pushed the updated version. This one initializes all entries in the merge table. Does it look more reliable?

DougGregor commented on lib/AST/LayoutConstraint.cpp:
```diff
@@ -143,53 +140,142 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
+    // If we got here, it means that the Self == Other check above has failed.
+    // And that could only happen if either the kinds are different or
+    // size/alignment parameters are different.
+    return LayoutConstraint::getUnknownLayout();
```
Okay, it probably doesn't matter, but one could merge exact-size(N) and at-most-size(N) into exact-size(N).

swiftix committed:
[GSB] Improve handling of layout constraints

This PR addresses TODOs from #8241.

- It supports merging for layout constraints, e.g., if both a _Trivial constraint and a _Trivial(64) constraint appear on a type parameter, we keep only _Trivial(64) as a more specific layout constraint. We do a similar thing for ref-counted/native-ref-counted. The overall idea is to keep the more specific of two compatible layout constraints.
- The presence of a superclass constraint implies a layout constraint, e.g., a superclass constraint implies _Class or _NativeClass

swiftix commented:
@swift-ci please smoke test

DougGregor commented:
Thank you!
